File: harness-engineering-py/backend/app/services/engine_factory.py
import asyncio
import json
import os
import time
from pathlib import Path
from typing import Any, Dict, List, Optional

from app.services.engine_interface import Engine
from app.services.opencode_wrapper import OpenCodeEngineWrapper
from app.models.schemas import EngineAvailability, ModelInfo


_engine_pool: Dict[str, Dict[str, Any]] = {}
ENGINE_POOL_TTL = 10 * 60  # 10 minutes in seconds

_cleanup_task_started = False


def _start_cleanup() -> None:
    global _cleanup_task_started
    if _cleanup_task_started:
        return
    _cleanup_task_started = True

    async def _cleanup_loop():
        while True:
            await asyncio.sleep(60)
            now = time.time()
            expired = [
                k for k, v in _engine_pool.items()
                if now - v["lastUsed"] > ENGINE_POOL_TTL
            ]
            for k in expired:
                entry = _engine_pool.pop(k, None)
                if entry:
                    try:
                        entry["engine"].cancel()
                    except Exception:
                        pass

    asyncio.ensure_future(_cleanup_loop())


async def get_or_create_engine(session_key: Optional[str] = None) -> Optional[Engine]:
    _start_cleanup()
    if session_key:
        cached = _engine_pool.get(session_key)
        if cached:
            cached["lastUsed"] = time.time()
            return cached["engine"]
    engine = await create_engine()
    if engine and session_key:
        _engine_pool[session_key] = {"engine": engine, "lastUsed": time.time()}
    return engine


async def create_engine() -> Optional[Engine]:
    engine = OpenCodeEngineWrapper()
    if not await engine.is_available():
        return None
    return engine


async def detect_engines() -> List[EngineAvailability]:
    engine = OpenCodeEngineWrapper()
    available = await engine.is_available()
    models = await _read_opencode_models()
    return [
        EngineAvailability(
            available=available,
            name="OpenCode",
            models=models,
            defaultModel=models[0].id if models else None,
        )
    ]


async def _read_opencode_models() -> List[ModelInfo]:
    config_path = Path.home() / ".config" / "opencode" / "opencode.json"
    if not config_path.exists():
        return [ModelInfo(id="default", name="Default")]

    try:
        content = config_path.read_text(encoding="utf-8")
        config = json.loads(content)
        models: List[ModelInfo] = []
        default_model = config.get("model", "")

        provider = config.get("provider", {})
        if isinstance(provider, dict):
            for provider_key, provider_val in provider.items():
                if not isinstance(provider_val, dict):
                    continue
                provider_models = provider_val.get("models", {})
                if not isinstance(provider_models, dict):
                    continue
                for model_key, model_val in provider_models.items():
                    if not isinstance(model_val, dict):
                        continue
                    full_id = f"{provider_key}/{model_key}"
                    is_default = default_model == full_id
                    models.append(ModelInfo(
                        id=full_id,
                        name=(model_val.get("name", model_key) + (" (默认)" if is_default else "")),
                    ))

        if not models:
            return [ModelInfo(id="default", name="Default")]

        # Sort: default model first, then alphabetically
        models.sort(key=lambda m: ("" if m.id == default_model else m.name))
        return models
    except Exception:
        return [ModelInfo(id="default", name="Default")]


async def is_engine_available() -> bool:
    engine = await create_engine()
    return engine is not None

File: harness-engineering/src/lib/engines/engine-factory.ts
import { OpenCodeEngineWrapper } from "./opencode-wrapper"
import type { Engine } from "./engine-interface"
import type { EngineAvailability, ModelInfo } from "@/types/chat"
import { readFile } from "fs/promises"
import { existsSync } from "fs"
import { resolve } from "path"
import { homedir } from "os"

export type EngineType = "opencode"

const enginePool = new Map<string, { engine: Engine; lastUsed: number }>()

setInterval(() => {
  const now = Date.now()
  for (const [key, entry] of enginePool) {
    if (now - entry.lastUsed > ENGINE_POOL_TTL) {
      if (typeof (entry.engine as any).cancel === "function") (entry.engine as any).cancel()
      enginePool.delete(key)
    }
  }
}, 60_000)

const ENGINE_POOL_TTL = 10 * 60 * 1000

export async function getOrCreateEngine(sessionKey?: string): Promise<Engine | null> {
  if (sessionKey) {
    const cached = enginePool.get(sessionKey)
    if (cached) {
      cached.lastUsed = Date.now()
      return cached.engine
    }
  }
  const engine = await createEngine()
  if (engine && sessionKey) enginePool.set(sessionKey, { engine, lastUsed: Date.now() })
  return engine
}

export async function createEngine(): Promise<Engine | null> {
  const engine = new OpenCodeEngineWrapper()
  if (!(await engine.isAvailable())) return null
  return engine
}

export async function detectEngines(): Promise<EngineAvailability[]> {
  const engine = new OpenCodeEngineWrapper()
  const available = await engine.isAvailable()
  const models = await readOpenCodeModels()
  return [{ available, name: "OpenCode", models, defaultModel: models[0]?.id }]
}

async function readOpenCodeModels(): Promise<ModelInfo[]> {
  const configPath = resolve(homedir(), ".config", "opencode", "opencode.json")
  if (!existsSync(configPath)) return [{ id: "default", name: "Default" }]

  try {
    const content = await readFile(configPath, "utf-8")
    const config = JSON.parse(content)
    const models: ModelInfo[] = []
    const defaultModel = config.model || ""

    if (config.provider && typeof config.provider === "object") {
      for (const [providerKey, providerVal] of Object.entries(config.provider)) {
        const provider = providerVal as { models?: Record<string, any> }
        if (!provider.models || typeof provider.models !== "object") continue
        for (const [modelKey, modelVal] of Object.entries(provider.models)) {
          const model = modelVal as { name?: string }
          const fullId = `${providerKey}/${modelKey}`
          const isDefault = defaultModel === fullId
          models.push({
            id: fullId,
            name: model.name || modelKey + (isDefault ? " (默认)" : ""),
          })
        }
      }
    }

    if (models.length === 0) return [{ id: "default", name: "Default" }]
    models.sort((a, b) => {
      if (a.id === defaultModel) return -1
      if (b.id === defaultModel) return 1
      return a.name.localeCompare(b.name)
    })
    return models
  } catch {
    return [{ id: "default", name: "Default" }]
  }
}

export async function isEngineAvailable(): Promise<boolean> {
  const engine = await createEngine()
  return engine !== null
}
